db/rdbms/src/main/java/io/camunda/db/rdbms/write/queue/ExecutionQueue.java

```diff
@@ -53,56 +55,100 @@ public void registerPostFlushListener(final PostFlushListener listener) {
+        return 0;
       }
-
-      if (!preFlushListeners.isEmpty()) {
-        LOG.debug("[RDBMS ExecutionQueue, Partition {}] Call pre flush listeners", partitionId);
-        preFlushListeners.forEach(PreFlushListener::onPreFlush);
+      LOG.debug(
+          "[RDBMS ExecutionQueue, Partition {}] Flushing execution queue with {} items",
+          partitionId,
+          queue.size());
+
+      final var startMillis = System.currentTimeMillis();
+      final var session =
+          sessionFactory.openSession(
+              ExecutorType.BATCH, TransactionIsolationLevel.READ_UNCOMMITTED);
+
+      var flushedElements = 0;
+      try {
+        while (!queue.isEmpty()) {
+          final var entry = queue.peek();
+          LOG.trace("[RDBMS ExecutionQueue, Partition {}] Executing entry: {}", partitionId, entry);
+          session.update(entry.statementId(), entry.parameter());
+          queue.remove();
+          flushedElements++;
+        }
+
+        if (!preFlushListeners.isEmpty()) {
+          LOG.debug("[RDBMS ExecutionQueue, Partition {}] Call pre flush listeners", partitionId);
+          preFlushListeners.forEach(PreFlushListener::onPreFlush);
+        }
+
+        session.flushStatements();
+        session.commit();
+        if (!postFlushListeners.isEmpty()) {
+          LOG.debug("[RDBMS ExecutionQueue, Partition {}] Call post flush listeners", partitionId);
+          postFlushListeners.forEach(PostFlushListener::onPostFlush);
+        }
+        LOG.debug(
+            "[RDBMS ExecutionQueue, Partition {}] Commit queue with {} entries in {}ms",
+            partitionId,
+            flushedElements,
+            System.currentTimeMillis() - startMillis);
+
+        return flushedElements;
+      } catch (final Exception e) {
+        LOG.error(
+            "[RDBMS ExecutionQueue, Partition {}] Error while executing queue", partitionId, e);
+        session.rollback();
+
+        throw e;
+      } finally {
+        session.close();
       }
+    }
+  }
 
-      session.flushStatements();
-      session.commit();
-      if (!postFlushListeners.isEmpty()) {
-        LOG.debug("[RDBMS ExecutionQueue, Partition {}] Call post flush listeners", partitionId);
-        postFlushListeners.forEach(PostFlushListener::onPostFlush);
+  /**
+   * Iterate from end over the queue and try to find a last added compatible queueItem. The
+   * queueItem will be replaced with a new, combined queueItem.
+   */
+  public boolean tryMergeWithExistingQueueItem(final QueueItemMerger... combiners) {
+    synchronized (queue) {
+      int index = queue.size() - 1;
+      for (final Iterator<QueueItem> it = queue.descendingIterator(); it.hasNext(); ) {
+        final QueueItem item = it.next();
+
+        for (final QueueItemMerger merger : combiners) {
+          if (merger.canBeMerged(item)) {
+            LOG.debug("Merging new item with item {}, {}", item.contextType(), item.id());
+            final var newItem = merger.merge(item);
```

Comment:
since newItem is just used once, this could also be inlined in the queue.set function.